controller/controller_client.go

```diff
@@ -5,284 +5,251 @@
 	_, level, unitConfig := unit.Expected()
+
 	cfg, err := config.ShipperConfigFromUnitConfig(level, unitConfig)
 	if err != nil {
-		c.log.Errorf("error unpacking config from agent: %s", err)
-		_ = unit.UpdateState(client.UnitStateFailed, err.Error(), nil)
+		c.reportError("error configuring shipper", err, unit)
 		return
 	}
 
-	err = logp.Configure(cfg.Log)
+	err = c.outputHandler.Start(cfg)
 	if err != nil {
-		c.log.Errorf("error unpacking config from agent: %s", err)
-		_ = unit.UpdateState(client.UnitStateFailed, err.Error(), nil)
+		c.reportError("error starting output server shipper", err, unit)
 		return
 	}
+	_ = unit.UpdateState(client.UnitStateHealthy, "outputs initialized", nil)
+}
+
+// called when we get a UnitUpdated for the output unit
+func (c *clientHandler) updateShipper(unit *client.Unit) {
+	c.log.Debugf("updating output unit %s", unit.ID())
+	state, _, _ := unit.Expected()
+	c.units.SetOutput(unit)
+	if state == client.UnitStateHealthy { // config update, so restart
+		_ = unit.UpdateState(client.UnitStateStopping, "shipper is restarting", nil)
+		c.stopShipper()
+		c.startShipper(unit)
+	} else if state == client.UnitStateStopped { // shut down
+		_ = unit.UpdateState(client.UnitStateStopping, "shipper is stopping", nil)
+		c.stopShipper()
+		_ = unit.UpdateState(client.UnitStateStopped, "shipper is stopped", nil)
+	}
+}
 
-	// reset the local logger
-	c.log = logp.L()
-
-	err = c.Run(cfg, unit)
+/*
+//////////
+////////// Input Handlers
+/////////
+*/
+
+func (c *clientHandler) startgRPC(unit *client.Unit, cfg config.ShipperClientConfig) {
+	//TODO: until we get TLS config fixed/figured out, run in insecure mode
+	// certPool := x509.NewCertPool()
+	// for _, cert := range cfg.Shipper.Server.TLS.CAs {
+	// 	if ok := certPool.AppendCertsFromPEM([]byte(cert)); !ok {
+	// 		c.reportError("error appending cert obtained from input in shipper startup", err, outUnit)
+	// 		return
+	// 	}
+	// }
+
+	creds := insecure.NewCredentials() //:= credentials.NewTLS(&tls.Config{
+	// 	ClientAuth:     tls.RequireAndVerifyClientCert,
+	// 	ClientCAs:      certPool,
+	// 	GetCertificate: c.getCertificate,
+	// 	MinVersion:     tls.VersionTLS12,
+	// })
+
+	err := c.grpcServer.Start(creds, cfg.Server)
 	if err != nil {
-		c.log.Errorf("error running shipper: %s", err)
-		_ = unit.UpdateState(client.UnitStateFailed, err.Error(), nil)
+		c.reportError("failed to start grpc server", err, unit)
 		return
 	}
+	c.log.Debugf("gRPC started")
+
+}
+
+func (c *clientHandler) stopGRPC() {
+	c.grpcServer.Stop()
+	c.log.Debugf("gRPC server stopped")
 }
 
 // start an individual input stream
-func (c *clientHandler) startStream(unit *client.Unit) {
-	// when we have individual input streams, that'll go here.
-	c.log.Debugf("Got unit stream for processor: %s", unit.ID())
+func (c *clientHandler) addInput(unit *client.Unit) {
+	// I assume that once we connect output processors, that will go somewhere here
```

Comment:
```suggestion
```

I don't know that we need this comment, since it doesn't explain anything about the current implementation. It is just speculating about some future implementation.